### ISSUE TITLE
Stutter and other Performance Concerns

### ISSUE THREAD

**Issue by Fugaz1** [Hekili/hekili#2624]
### If adding to the thread, you must:
* Include a current snapshot taken when the issue was occurring.
* Report the last version of the addon that did not have this issue for you.  You can downgrade your addon version to investigate further.
* Note any additional observations about when/where this occurs (or if it improves/worsens in different settings).
* After testing in combat, `/dump HekiliEngine.threadUpdates` and copy/paste the text (or a screenshot) showing the output.

### If the above is missing from your post, I will probably delete it.  "I'm having this issue, too" is not useful for troubleshooting the issue.  Thanks for your understanding.

Back to the original ticket...

### Before You Begin

- [X] I confirm that I have downloaded the latest version of the addon.
- [X] I am not playing on a private server.
- [X] I checked for an [existing, open ticket](https://github.com/Hekili/hekili/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc) for this issue and was not able to find one.
- [X] I edited the title of this bug report (above) so that it describes the issue I am reporting.

### Describe the Issue

Hardware: 7800X3D, 4070ti
Addons: only Hekili [v10.1.0-1.0.1c]

I noticed there is constant microstutter when hekili is running, in and out of combat. Pausing hekili fixes the issue.

Here is comparision, it is in orgrimmar and I am looking at wall so nothing is going on, all addons but hekili are disabled.
1) This is without hekili:
![image](https://github.com/Hekili/hekili/assets/759426/a25d1ed7-3482-49d7-bdde-fce47c7ee599)
2) This one is with hekili:
![image](https://github.com/Hekili/hekili/assets/759426/68c5cd73-61d6-4dff-8d58-d86e651cf876)

This is with vengeance DH, with havoc 1% low goes down to 55. I also tried evoker - it was closer to vengeance.

In combat this problem amplifies, example for target dummy with havoc DH: 1% low goes from 100 to 40 (!) FPS. When I reset my profile it goes to 55 (strange as I have 2 abilities in queue and default is 3).

I tried similar addons and has no such stutter with maxdps or conro.

### How to Reproduce

Load hekili, check 1% lows with RTSS - no need for combat or anything.

### Player Information (Link)

https://pastebin.com/AYkKtNKX

### Error Messages (Link)

no errors

**Comment by Fugaz1** [comment]
P.S. I tried to play with performance options but it didn't make any positive change.

**Comment by Hekili** [comment]
> I noticed there is constant microstutter when hekili is running, in and out of combat. Pausing hekili fixes the issue.

There is a known issue that I am actively investigating.  I have not found pausing to actually resolve it, for me.

From your snapshot:
```
    maxTime = 100
    ...
    throttleRefresh = false
    throttleTime = true
```

These settings mean that you've told the addon to override its default limitation of ~15ms of calculations in a frame, allowing up to 100.  You may have done this at my instruction, since one current theory is that splitting the work over multiple frames ends up having a larger performance impact than just blasting out new recommendations in a single thread.  If you uncheck "Set Update Time" in your spec's Performance options, is there any notable difference in performance?

What does this mean?  1% low?
> This is with vengeance DH, with havoc 1% low goes down to 55. I also tried evoker - it was closer to vengeance.

The framerate or CPU usage testing should be broken down more like the following grid:

 x | In Combat | Out of Combat
--|-|-
Hekili Loaded + Enabled | | 
Hekili Loaded + Disabled | |
Hekili Not Loaded | |

For example, I don't know how much your framerate is impacted entering combat without the addon, if at all.

> When I reset my profile it goes to 55 (strange as I have 2 abilities in queue and default is 3).

This is *probably* due to your Performance tab settings getting reset, but I'm unsure.  Resetting your profile shouldn't be necessary, regardless.

**Comment by Fugaz1** [comment]
I tested different performance settings, not much difference there. Biggest difference in 1% lows was changing specs, so for example vengance 75, havoc 55.

For me pausing was same as not loading addon. Unpausing caused stutter no matter if I was standing if middle of nowhere or was fighting. Obviously fighting made things worse (havoc 55 -> 40 fps 1%). So table would look like this:

x | In Combat | Out of Combat
-- | -- | --
Hekili Loaded + Enabled |  stutter |  stutter 
Hekili Loaded + Disabled |  OK |  OK
Hekili Not Loaded |  OK |  OK 

1% low is like minimum fps but over time, basically it measures drops and stutters. What it tells us here for example is when hekili is doing update my fps drops from 138 to 75 which is clearly visible in RTTS graphs (one is 1% low and another is render latency).

You can install MSI Afterburner and enable OSD there (it uses RTTS):
![image](https://github.com/Hekili/hekili/assets/759426/89aa2e70-9281-45df-b66f-f6386b3d65eb)
Hotkeys to toggle OSD and start/stop fps benchmark are in OSD and Benchmark tabs.
With this tool you easily spot and check what is causing it.

**Comment by felipegaraujo** [comment]
Also having microstutters.

https://pastebin.com/gbcsY9Xq

```Dump: value=HekiliEngine.threadUpdates 
[1]={ 
  meanClockTime=32.931264332262, 
  meanWorkTime=22.998234527653, 
  totalWasted=75.663800016046, 
  peakWasted=15.523400001228, 
  peakFrames=3, 
  peakWorkTime=40.568700000644, 
  updates=614, 
  firstUpdate=36186133.3207, 
  updatesPerSec=1.6995977619553, 
  peakClockTime=60.165399998426, 
  meanWasted=0.80310434858212, 
  meanFrames=2.0798045602606 
}
```

![off](https://github.com/Hekili/hekili/assets/6819753/eedd2cea-aeb4-4c23-b68f-18e6983d31b7) - hekili disabled
![on](https://github.com/Hekili/hekili/assets/6819753/bc5efb09-0f63-4138-8b5d-8ccadd39ac1c) - hekili enabled


Always had this issue, just tried to ignore the microstutters and moved on.

Hardware 3700x and GTX 1070.

Hope this helps figuring it out.

**Comment by Cristisnyke** [comment]
with a 13900k and a 4090 i too can reproduce this type of behaviour - it seems to depend heavily upon wether or not "Enhanced recheck" is used for a particular spec - for example on my outlaw rogue - at training dummy in valdraken i get around 150 fps

Turn off enhanced recheck and reload the game - suddenly i get around 180.

Its a wierd behaviour and its mostly noticeable on high refresh rate monitors - where the slightest stutter can be noticed.

I would say the microstutter is there every time the addon makes a call and needs to recalculate.

Depending on your settings of course you can agravate or aleviate the issues. 

This is outlaw with enhanced recheck on - default settings for refresh rate and throttle time. 

![image](https://github.com/Hekili/hekili/assets/73036971/1b4744c6-dcb0-46fe-8928-3af94f66dbc5)

We can agravate the issue by asking the addon to update more often.

For example with a 0.05 refreshrate this is what you can expect.

![image](https://github.com/Hekili/hekili/assets/73036971/2493359a-c0d1-4186-8735-fe77582fd825)

All addons disabled Except Hekili with enhanced recheck on and default refresh rate.
![image](https://github.com/Hekili/hekili/assets/73036971/fabffcc6-dd48-460d-ba27-354978b76d62)

Addons disabled just for memes
![image](https://github.com/Hekili/hekili/assets/73036971/3d219d5b-cf00-44a4-bba2-d35e26dafe9c)

The problem while yes is partly caused by Hekili.
I do not think any amount of optimization will be able to fix it.

The addons share the same thread as the main Gx engine - they are competing for the same resources - the same CPU thread
This is a very old game engine that still to this day even with Dx12 is not capable of utilizing system resources properly.

Untill they manage to split addons to run on a different cpu thread this problem will remain.

My advice is - Find the minimum possible number of addons to run along side hekili. It will be way less noticeable.

Weakauras are the biggest offenders and nameplate addons tend to be as well.

See if you can survive just running hekili and a bossmod.
Things that dont do stuff constantly - IE Leatrix plus - Junk sellers - Most quality of life addons shouldn't be a problem.

**Comment by codengine** [comment]
Just a thought: I know that Hekili uses the SimC rotations for evaluating the suggestions. In case that the rotations are parsed on the fly, would it be possible to have some kind of preprocessor that generates an optimized "rotation file" which gets rid of some of the overhead?  Like, native LUA code for example?

Have you checked a dominator graph which routines consume which time? My performance is also dunking heavily in raids.

Moreover, could we have some kind of a simple model where only the very next action is recommended and no prediction for the future is happening or would that ruin the results?

**Comment by Dragonfo** [comment]
The problem lies in the addon itself, not in the game and the game engine. Why I say so, there is an analogue of Hekili is HeroRotation. Which acts as an assistant in the rotation of the game. If you consider HeroRotation you may notice a couple of things. The addon does not load the CPU in M+ in raids and in the open world. Clicking on abilities is fast and clear without the 1-2 sec delay like Hekili. There is also a handy feature that displays abilities on nameplate opponents, which would seem to also load the CPU, but no. 
I suppose that we need to review the very structure of the addon Hekili as a whole and simplify some things that most players do not even use but in the addon are present. I would just recommend familiarizing yourself with HeroRotation's structure and borrowing their optimization and addon performance work.

Translated with www.DeepL.com/Translator (free version)

**Comment by Cristisnyke** [comment]
> The problem lies in the addon itself, not in the game and the game engine. Why I say so, there is an analogue of Hekili is HeroRotation. Which acts as an assistant in the rotation of the game. If you consider HeroRotation you may notice a couple of things. The addon does not load the CPU in M+ in raids and in the open world. Clicking on abilities is fast and clear without the 1-2 sec delay like Hekili. There is also a handy feature that displays abilities on nameplate opponents, which would seem to also load the CPU, but no. I suppose that we need to review the very structure of the addon Hekili as a whole and simplify some things that most players do not even use but in the addon are present. I would just recommend familiarizing yourself with HeroRotation's structure and borrowing their optimization and addon performance work.

Somehow i doubt its that simple. Those addons are not as accurate as hekili and lack 3/4ths of the available features.

**Comment by Dragonfo** [comment]
> > Проблема кроется в самом аддоне, а не в игре и игровом движке. Почему я так говорю, есть аналог Hekili это HeroRotation. Который выступает помощником в ротации игры. Если вы рассмотрите HeroRotation, вы можете заметить пару вещей. Аддон не нагружает процессор в М+ в рейдах и в открытом мире. Щелчок по способностям быстрый и четкий, без задержки в 1-2 секунды, как у Хекили. Еще есть удобная фича, выводящая способности на именных противников, что вроде бы тоже нагружает ЦП, но нет. Я предполагаю, что нам нужно пересмотреть саму структуру аддона Hekili в целом и упростить некоторые вещи, которые большинство игроков даже не используют, но в аддоне присутствуют. Я бы просто рекомендовал ознакомиться со структурой HeroRotation и позаимствовать их работу по оптимизации и производительности аддонов.
> 
> Что-то я сомневаюсь, что это так просто. Эти аддоны не так точны, как hekili, и им не хватает 3/4 доступных функций.
I've been using both Hekili and HeroRotation for a long time. And I can confirm it with a screenshot of logs. With HeroRotation, my buttons and rotations are more responsive than with Hekili. Accuracy isn't always good when one addon does 80%+ of the loadouts of your interface in the game. Plus I as described above there is a lag in clicking with Hekili. Just an example take a melee class, and try to quickly spam the rotation keys and see that by quickly pressing keys with Hekili you will have latency. with HeroRotation you won't have latency, you'll just spam as fast as you can. Here is the first screenshot where I have all purple orange logs, this is with HeroRotation. The second screenshot is from Hekili. I have the logs below on the alt. Because there is a delay on CastDelay there is a delay in the rotation update itself at Hekili and there is a problem with my i5-12400 GT1650 going down in fps. On this I did not just point out and said that the problem in the engine addon Hekili and if you make the optimization and performance as HeroRotation. Then I think it will be the best addon. But so far there are delays in updating rotation delays when you press keys, and lower fps. It's not playable with Hekili. Anyone who wants to go install HeroRotation, it doesn't have most of the rotation support, but the ones that do work fine. 

![image](https://github.com/Hekili/hekili/assets/63850853/bff82c93-3d08-4c63-b705-ef1c0e7ef82b)
![image](https://github.com/Hekili/hekili/assets/63850853/bd8be675-3c2f-433c-abf5-95450b67c109)

**Comment by Hekili** [comment]
Just a reminder, there are requirements for posting in this issue thread in the first post. I'll be flexible to an extent, but don't be surprised if your posts get deleted if you're not supplying information for investigating the topic. 

A few folks have mentioned that there are older versions that do/don't have the stuttering for them. If that's true for you, adding information about the last version that does not have the stutter or the first version that does would be helpful. 

Also, this addon does use more CPU than average and there are definitely opportunities to improve efficiency. For systems with a very high framerate, I would have to throttle the addon's CPU usage per frame. This technically exists already (Performance tab) but it's not intuitive to use.

**Comment by Dragonfo** [comment]
![image](https://github.com/Hekili/hekili/assets/63850853/6b133b97-04bc-4a1c-9c9f-5b0daeffdfea)
![image](https://github.com/Hekili/hekili/assets/63850853/a42dd136-0448-4d2a-a226-93e9368a1b7f)
Here are two screenshots. The first at idle and in Hekili load.
![image](https://github.com/Hekili/hekili/assets/63850853/d8f8aa9f-6f95-4b44-8c5b-d33acdbfb04b)
![image](https://github.com/Hekili/hekili/assets/63850853/b2464809-70e4-4015-9db6-ed2a670961cc)
The second for the sake of pure comparison HeroRotation. Well, as if there is a tangible difference. And that's why such a difference between the two addons? at idle they are the same there + - error, but in the load there fps drops just strongly. My components are i5-12400f gt1650.

**Comment by Hekili** [comment]
I'm doing some testing with a dynamic throttle that will change how much time the addon can use per frame depending on your framerate.

The logic for this is:
* When the addon is *not* generating recommendations, track how much time elapses for each frame.
* When it is time to generate new recommendations, calculate the average time-per-frame from the above.
* Allow the update thread to use 60% of that average time.

This means that, if your system has 200 FPS (which is wild, I'm hypothetically jealous), your average time-per-frame is 5ms, and the addon would allow for 3ms (5ms \* 60%) per frame while updating recommendations.  If your framerate is 60 FPS, that's 16.667ms, so the addon would allow you for 10ms (16.667ms \* 60%) per frame.

I currently have constrained the allowed time-per-frame to between 3ms and 10ms.  If your framerate goes below 60 FPS (which seems common in combat), I don't want the addon's CPU appetite to grow, impacting your framerate, which then could budget even more time per frame in a vicious cycle.  However, if someone gets 10 FPS baseline -- 100ms per frame -- it could backfire in terms of responsiveness if the addon leaves 90% of the available CPU time untouched.

This has worked really well for my Demonology Warlock, but I'm seeing some wasted effort on my Enhancement Shaman.  For Enhancement, I *think* the issue is that the addon can also get forced to restart its recommendation calculations because of Maelstrom Weapon proccing randomly and because some auras can pop-up after a delay.  I'm assessing what I can do about it now.  I'm considering forcing the addon to wait a bit longer after critical events before starting to generate new recommendations, but that could also feel sluggish.

Future considerations:
* Revise the 'reset' process to reduce risk of rework.
* Update auras incrementally using modern aura events instead of flagging for a complete purge/refresh.
* Flatten the process of generating recommendations to reduce recursion/loops/nesting and forego unnecessary throwaway work.

**Comment by codengine** [comment]
@Hekili Happy to gather some data. Would it be an idea that you could provide a profiling version of the addon with additional ways of measurements? Or does the snapshot already contain all that you need?

**Comment by Hekili** [comment]
Thanks; I'll post an alpha build 'soon' (next few days).  Last night's testing wasn't super promising, to be quite honest.

**Comment by Hekili** [comment]
@codengine I've posted the initial version of the recommendation throttling that works as described.  I'm going to expose more of the tuning knobs for it in the addon's options.  After combat, you can supply the output from `/dump HekiliEngine.threadUpdates` along with any performance differences that you notice (better or worse).

**Comment by cremor** [comment]
I never had performance problems, but maybe it helps anyway. This is after an M+, for a Ret Paladin:

```
Dump: value=HekiliEngine.threadUpdates
[1]={
  meanClockTime=14.406355861171,
  meanWorkTime=6.8003389973385,
  totalWasted=46.785799980164,
  peakWasted=27.937200069427,
  peakFrames=5,
  peakWorkTime=33.210099935532,
  updates=5562,
  firstUpdate=691311225.4296,
  updatesPerSec=1.506189064465,
  peakClockTime=92.972699999809,
  meanWasted=0.10425626962527,
  meanFrames=1.569219705142
}

```

**Comment by Breadwinka** [comment]
Heres after a raid encounter. I have noticed a lot less stuttering so far, and performance has been a lot better.
CPU 5800X3D
GPU 3080

```
Dump: value=HekiliEngine.threadUpdates 
[1]={ 
  meanClockTime=75.150609796957, 
  meanWorkTime=25.443050092296, 
  totalWasted=5427.5945999981, 
  peakWasted=32.27310000034, 
  peakFrames=9, 
  peakWorkTime=36.68730000034, 
  updates=7043, 
  firstUpdate=1522216.083, 
  updatesPerSec=1.0409564989257, 
  peakClockTime=198.33560000081, 
  meanWasted=5.7499321234801, 
  meanFrames=4.498935112878 
} 
```

**Comment by geraint425** [comment]
On Enhancement Shaman there is a delay in ability presses, especially when Ascendance ability is active and you need to press Windstrike often.
CPU i5-12400F
GPU 1650
```
Dump: value=HekiliEngine.threadUpdates
[1]={
  meanClockTime=39.075365225568,
  meanWorkTime=11.458779511277,
  totalWasted=206.26290000114,
  peakWasted=17.631599999964,
  peakFrames=7,
  peakWorkTime=22.75820000004,
  updates=1596,
  firstUpdate=1892386.9996,
  updatesPerSec=2.572071267253,
  peakClockTime=79.260400000028,
  meanWasted=1.7209048765124,
  meanFrames=4.2412280701754
}
```

**Comment by Hekili** [comment]
### If adding to the thread, you must:
* Include a current snapshot taken when the issue was occurring.
* Report the last version of the addon that did not have this issue for you.  You can downgrade your addon version to investigate further.
* Note any additional observations about when/where this occurs (or if it improves/worsens in different settings).
* After testing in combat, `/dump HekiliEngine.threadUpdates` and copy/paste the text (or a screenshot) showing the output.

### If the above is missing from your post, I will probably delete it.  "I'm having this issue, too" is not useful for troubleshooting the issue.  Thanks for your understanding.

**Comment by durrburger93** [comment]
Adding to this thread as I've been having this issue since the very start of DF, and I never used the addon before so idk if any previous version was exempt from constant stuttering.

The game micro stutters constantly during combat and it seems that every individual stutter occurs when I hit a spell and the addon loads up a new spell for the rotation. This has so far been happening on all of my characters so a warrior, fury/arms, havoc dh, outlaw rogue. 

Hardware shouldn't be the issue since I have a 4080/5800x3d/32gb of 3600mhz RAM/NVMe. Also I'm not sure how to get a snapshot since it wouldn't generate any after pausing and unpausing the addon.

Here are the screens on how the frame graph looks outside combat (hekili set to only show up in combat):

![no hekili](https://github.com/Hekili/hekili/assets/82004352/7f1ba7ce-dcf7-4377-aec8-9888160cdc99)

In combat with Havoc DH:

![hekili DH](https://github.com/Hekili/hekili/assets/82004352/0d6d810b-4067-4047-97c9-9327947a2a49)

In combat with Fury Warr:

![fury hekili](https://github.com/Hekili/hekili/assets/82004352/f1acd764-ae9f-40bf-a9d8-77a4a68d2fdb)

The output: 

![dump](https://github.com/Hekili/hekili/assets/82004352/7cb136f3-4b02-4781-9fd5-27db0bdd79ce)

**Comment by Hekili** [comment]
A snapshot is required to diagnose this issue.

## Snapshot

Please supply a Snapshot of the addon's decision-making when you are seeing this issue in-game. This is not a screenshot. To generate a Snapshot, please complete the following steps.

 * Log into your WoW character.
 * Recreate the issue you are reporting (i.e., generate 5 Holy Power).
 * When you see the recommendation you disagree with, press ALT-SHIFT-P to Pause and Snapshot (or use ALT-SHIFT-[ to Snapshot without pausing).
   * You can change these keybindings in /hekili > Toggles if needed.
   * If you take the snapshot at a different time, it will not be useful and you'll be asked to submit a new one.
 * Type /hekili and press Enter.
 * Open the Snapshots section on the left side.
 * Select the snapshot that shows the recommendation that you disagree with.
   * Snapshots are labeled with your specialization, the display in question (usually, you'll choose Primary), and the recommendations the addon made with their timings.
 * Click in the Export Snapshot box, press CTRL+A to select all text including hidden text and press CTRL+C to copy.
 * Open [https://pastebin.com](https://github.com/Hekili/hekili/issues/Pastebin), use CTRL+V to paste the snapshot text, then click Create New Paste.
 * Provide the Pastebin ***link*** in a new comment.

This step is essential, as most issues are related to specific classes, specializations, gear, talent choices, or other game systems.  If you do not provide this information, I cannot triage your problem.

**Comment by durrburger93** [comment]
Thanks but no snapshots are being generated, nothing happens when I press ALT-SHIFT-[ and the snapshot selection is grayed out even though ALT-SHIFT-[ is also the hotkey for me. I'm clicking the shortcut both while fighting a mob and after leaving combat, but nothing is happening.

**Comment by Hekili** [comment]
Use ALT+SHIFT+P instead.  Post screenshots of your snapshots keybindings.  Get BugSack/BugGrabber if you're pressing the button and not getting a snapshot.

**Comment by Tzuyume** [comment]
Dump: value=HekiliEngine.threadUpdates
[1]={
  meanClockTime=74.265713728762,
  meanWorkTime=19.909252033755,
  totalWasted=296.454599999889,
  peakWasted=21.443299997598,
  peakFrames=5,
  peakWorkTime=23.982599999756,
  updates=590,
  firstUpdate=16962271.097,
  updatesPerSec=2.9033932783827,
  peakClockTime=180.86590000018,
  meanWasted=2.8191393370451,
  meanFrames=4.032203389305
}
https://pastebin.com/ca67T9ax

This is one example of a spec dropping 10-20 fps during combat in the new raid. Fps drop doesn't happen at all in same situation with hekili off.

**Comment by Hekili** [comment]
> Dump: value=HekiliEngine.threadUpdates [1]={ meanClockTime=74.265713728762, meanWorkTime=19.909252033755, totalWasted=296.454599999889, peakWasted=21.443299997598, peakFrames=5, peakWorkTime=23.982599999756, updates=590, firstUpdate=16962271.097, updatesPerSec=2.9033932783827, peakClockTime=180.86590000018, meanWasted=2.8191393370451, meanFrames=4.032203389305 } https://pastebin.com/ca67T9ax
> 
> This is one example of a spec dropping 10-20 fps during combat in the new raid. Fps drop doesn't happen at all in same situation with hekili off.

Nothing significant has changed with Marksmanship.  There are video driver issues that caused freezes for me.  Let me know if you have an nVidia card and your drivers are up to date.

**Comment by Tazatal** [comment]
https://pastebin.com/UtiFqn2z

![338c92adcc88ad9ed77029cd65474f6f](https://github.com/Hekili/hekili/assets/124478480/8d51e837-5b19-4360-81d3-09a6682d3026)

Consistantly getting lag spikes within in raid during Lust and at start of fights and its not consistent every time for how long. On current Nvidia patch, and have had issues since start of 10.2 (CPU Profiler pointed to this having an issue)

Specs for comp if that helps.
CPU: Intel Core i9 9900K @ 3.60GHz
GPU: NVIDIA GeForce RTX 3060 Ti (Gigabyte)

**Comment by Hekili** [comment]
What version of nVidia drivers do you have?  As noted above, there was a bad release that caused performance issues.

I do plan to revise the way the addon queues its work in the near future, as I suspect that there may be an issue of video drivers combined with the coroutine usage of my addon, but it's hard to prove out.

**Comment by ploink-git** [comment]
![CleanShot 2023-12-11 at 20 07 57@2x](https://github.com/Hekili/hekili/assets/3453185/d34db037-e662-4e23-a84b-4d314829b43e)
Huge lag spikes when reloading and sometimes when finishing an m+ dungeon or pulling in raid

Maxed out 16" M3 Macbook Pro Max

**Comment by Tazatal** [comment]
> What version of nVidia drivers do you have? As noted above, there was a bad release that caused performance issues.
> 
> I do plan to revise the way the addon queues its work in the near future, as I suspect that there may be an issue of video drivers combined with the coroutine usage of my addon, but it's hard to prove out.

546.29 which is the latest version released on the 4th of Dec

**Comment by Jurandes** [comment]
Hello everyone, I Just noted one issue and probably its something wrong with my settings, but I noticed on my enh, that hekili is horribly wasting maelstrom weapon stacks which results in much less lb/EB instants.
Sometimes, ive noticed even having 10 stacks of MW, and primordial WAVE gettin off cd, IT tells me to cast pw, ergo - waste 10 free stacks...
Dont know What can I do with that, any advice?
Already switched off filler with fire brand but IT didnt help much. And suddenly performing worse double checking if to spend current MW stacks or to cast WAVE as IT shows.

**Comment by syrifgit** [comment]
> Hello everyone, I Just noted one issue and probably its something wrong with my settings, but I noticed on my enh, that hekili is horribly wasting maelstrom weapon stacks which results in much less lb/EB instants. Sometimes, ive noticed even having 10 stacks of MW, and primordial WAVE gettin off cd, IT tells me to cast pw, ergo - waste 10 free stacks... Dont know What can I do with that, any advice? Already switched off filler with fire brand but IT didnt help much. And suddenly performing worse double checking if to spend current MW stacks or to cast WAVE as IT shows.

If you have a specific issue you should open a new ticket and supply the required info (sims links, snapshots, etc) instead of tagging onto an unrelated issue.

**Comment by JLC827** [comment]
Hello, I am adding my own profiling data to this thread. Hekili is causing a significant drop in my 1% low FPS during combat, which is causing visual stuttering.

### 1. Observations and Profiling Data

I conducted a targeted performance test by running the exact same combat scenario twice: once with all addons disabled and once with only Hekili enabled.

The test involved attacking a **"Dungeoneer's Training Dummy"** at the **Contender's Gate in Dornogal**. The spec was holy priest.

The data below shows the direct impact of Hekili during combat:

| Addons Enabled | Average FPS | 1% Low FPS |
| :--- | :---: | :---: |
| **No Addons** | 105 | 68 |
| **Hekili Only** | 95 | 33 |

Under identical combat conditions, enabling only Hekili caused my 1% low FPS to plummet from 68 to 33, causing noticeable stuttering.

### 2. System and Addon Information

*   **Hardware:**
    *   **CPU:** AMD Ryzen 5 3600 6-Core Processor
    *   **GPU:** NVIDIA GeForce GTX 1660 SUPER
    *   **Driver:** NVIDIA Version 580.88
    *   **RAM:** 16 GB
    *   **Motherboard:** B450M PRO-VDH MAX (MS-7A38)
*   **Hekili Version:** `v11.2.0-1.0.0f`
*   **Last Good Version:** I do not have data for a previous version, as this is the first time I have performed this level of profiling.

### 3. Combat Test: `/dump HekiliEngine.threadUpdates`

The following diagnostic data was collected during the combat test with Hekili enabled.

```
Dump: value=HekiliEngine.threadUpdates
[1]={
  meanClockTime=15.571152631586,
  meanFrameTime=11.817816256379,
  meanWorkTime=14.872062781966,
  totalWasted=0,
  peakWasted=0,
  peakWorkTime=21.841499999166,
  peakFrames=2,
  peakClockTime=37.391300000025,
  updates=798,
  firstUpdate=15244677.1109,
  updatesPerSec=1.6090980880441,
  framesWorked=205550,
  meanWasted=0,
  meanFrames=1.0576441102757
}
```

### 4. Hekili Snapshot

A snapshot containing my settings, talents, and gear was taken immediately after the combat test. The full snapshot string is available via the Pastebin link below.

[https://pastebin.com/ttguVxrf](https://pastebin.com/ttguVxrf)

Thank you for your work on this addon. I hope this direct combat comparison data is useful for diagnosing and resolving the performance bottleneck. Please let me know if any other information would be helpful.

**Comment by syrifgit** [comment]
Addon performance has been completely overhauled since this ticket, it would be more appropriate to open a new ticket instead of adding to a very old one.

And make sure you test somewher besides dornogal which has horrible stuttering and lag for everyone.
